src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1490 @@
+    }
+
+    @Override
+    public void visitIdent(JCIdent tree) {
+
+        // Reference to this?
+        if (tree.name == names._this || tree.name == names._super) {
+            refs.mapInto(refs, ThisRef.class, direct -> new ExprRef(depth, direct));
+            return;
+        }
+
+        // Parameter or local variable?
+        if (isParamOrVar(tree.sym)) {
+            VarSymbol sym = (VarSymbol)tree.sym;
+            if (refs.contains(VarRef.direct(sym)))
+                refs.add(ExprRef.direct(depth));
+            if (refs.contains(VarRef.indirect(sym)))
+                refs.add(ExprRef.indirect(depth));
+            return;
+        }
+
+        // An unqualified, non-static method invocation must reference 'this' or outer 'this'.
+        // The "value" of a non-static method is a reference to its instance.
+        if (tree.sym.kind == MTH && (tree.sym.flags() & Flags.STATIC) == 0) {
+            MethodSymbol sym = (MethodSymbol)tree.sym;
+
+            // Check for implicit 'this' reference
+            ClassSymbol methodClassSym = methodClass.sym;
+            if (methodClassSym.isSubClass(sym.owner, types)) {
+                refs.mapInto(refs, ThisRef.class, direct -> new ExprRef(depth, direct));
+                return;
+            }
+
+            // Check for implicit outer 'this' reference
+            if (methodClassSym.isEnclosedBy((ClassSymbol)sym.owner)) {
+                refs.mapInto(refs, OuterRef.class, direct -> new ExprRef(depth, direct));
+                return;
+            }
+
+            // What could it be?
+            //Assert.check(false);
+            return;
+        }
+
+        // Unknown
+        return;
+    }
+
+    @Override
+    public void visitSynchronized(JCSynchronized tree) {
+        scan(tree.lock);
+        refs.discardExprs(depth);
+        scan(tree.body);
+    }
+
+    @Override
+    public void visitConditional(JCConditional tree) {
+        scan(tree.cond);
+        refs.discardExprs(depth);
+        RefSet<ExprRef> combinedRefs = new RefSet<>();
+        scan(tree.truepart);
+        combinedRefs.addAll(refs.removeExprs(depth));
+        scan(tree.falsepart);
+        combinedRefs.addAll(refs.removeExprs(depth));
+        refs.addAll(combinedRefs);
+    }
+
+    @Override
+    public void visitIf(JCIf tree) {
+        scan(tree.cond);
+        refs.discardExprs(depth);
+        scan(tree.thenpart);
+        scan(tree.elsepart);
+    }
+
+    @Override
+    public void visitExec(JCExpressionStatement tree) {
+        scan(tree.expr);
+        refs.discardExprs(depth);
+    }
+
+    @Override
+    public void visitThrow(JCThrow tree) {
+        scan(tree.expr);
+        if (refs.discardExprs(depth))     // we don't try to "catch" refs from thrown exceptions
+            leakAt(tree);
+    }
+
+    @Override
+    public void visitAssert(JCAssert tree) {
+        scan(tree.cond);
+        refs.discardExprs(depth);
```

Comment:
Should also handle `tree.detail`? (Obscure of course, but could be `assert true : foo(this);`.)

Reply:
Thanks - will fix.